### PR TITLE
8296787: Unify debug printing format of X.509 cert serial numbers

### PR DESCRIPTION
**Pull request by blperez01**
Updated `X509CertSelector.java`, `SerialNumber.java`, and `SSLLogger`  to use same debug printing format for X.509 serial numbers. The changes are:
-  `X509CertSelector` was `SN: 83be056904246b1a1756ac95991c74a` and now is `Serial number: 08:3B:E0:56:90:42:46:B1:A1:75:6A:C9:59:91:C7:4A`
- `SerialNumber` was `SerialNumber: [ 083be056 904246b1 a1756ac9 5991c74a]` and now is `Serial number: 08:3B:E0:56:90:42:46:B1:A1:75:6A:C9:59:91:C7:4A`
- `SSLLogger` was `"serial number" : "083BE056904246B1A1756AC95991C74A"` and now is `"serial number" : 08:3B:E0:56:90:42:46:B1:A1:75:6A:C9:59:91:C7:4A`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296787](https://bugs.openjdk.org/browse/JDK-8296787): Unify debug printing format of X.509 cert serial numbers (**Bug** - P4)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**) ⚠️ Review applies to [2b6b2c9f](https://git.openjdk.org/jdk/pull/16834/files/2b6b2c9f11aea4ff17adf0d0df1769b497f3a935)
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**) ⚠️ Review applies to [2b6b2c9f](https://git.openjdk.org/jdk/pull/16834/files/2b6b2c9f11aea4ff17adf0d0df1769b497f3a935)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16834/head:pull/16834` \
`$ git checkout pull/16834`

Update a local copy of the PR: \
`$ git checkout pull/16834` \
`$ git pull https://git.openjdk.org/jdk.git pull/16834/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16834`

View PR using the GUI difftool: \
`$ git pr show -t 16834`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16834.diff">https://git.openjdk.org/jdk/pull/16834.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16834#issuecomment-1828677545)

### CODE REVIEW

**Comment by coffeys** [comment]
for clarity - can you include before and after examples of the serial numbers printed from each change ?

**Comment by seanjmullan** [comment]
I would like us to use a consistent hex format where each byte is separated by a ":", so for example:

08:3B:E0:56:90:42:46:B1:A1:75:6A:C9:59:91:C7:4A

This is easier to visually parse and is a common format used by other X.509 certificate display tools.

File: src/java.base/share/classes/sun/security/x509/SerialNumber.java [hunk 1]
```diff
@@ -100,9 +100,8 @@ public SerialNumber(InputStream in) throws IOException {
     /**
      * Return the SerialNumber as user readable string.
      */
-    public String toString() {
-        return "SerialNumber: [" + Debug.toHexString(serialNum) + ']';
-    }
+    public String toString() { return "Serial number: " +
+        serialNum.toString(16).toUpperCase().replaceAll("(?<=..)(..)", ":$1"); }
```
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/x509/SerialNumber.java:104]: I am not sure that `replaceAll` expression is correct. When I tried it for `BigInteger("74984748")`, it emitted "47:82:d2c" when I think it should be "04:78:2D:2C". I would use the HexFormat class here (and in `X509CertSelector`), ex:

`HexFormat.ofDelimiter(":").withUpperCase().formatHex(xcert.getSerialNumber().toByteArray())`

**Comment by seanjmullan** [comment]
> for clarity - can you include before and after examples of the serial numbers printed from each change ?

Yes, I think that would be useful. Also you will need to add a `noreg` label to the bug since you are not including a test (which I think is ok).

**Review by seanjmullan — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/sun/security/x509/SerialNumber.java [hunk 2]
```diff
@@ -100,9 +101,8 @@ public SerialNumber(InputStream in) throws IOException {
     /**
      * Return the SerialNumber as user readable string.
      */
-    public String toString() {
-        return "SerialNumber: [" + Debug.toHexString(serialNum) + ']';
-    }
+    public String toString() { return "Serial number: " +
```
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/x509/SerialNumber.java:104]: Here I would keep "Serial number" as "SerialNumber" to maintain consistency with other X.509 `toString` methods which generally emit the name of the field using camel case format.

My only other comment is to whether it makes sense to move this into a utility method. I see `sun.security.util.Debug.toString` already does the same thing and uses HexFormat but in lower case. I think using lower case or upper case is not a big deal. So maybe just change all of these code instances to call `Debug.toString` to keep things consistent and reuse existing code.

**Comment by coffeys** [comment]
Edits are heading in the right direction. 

How about refactoring the Debug utility methods ? `toString(byte[]) `might be better as `toHexString(byte[])`

you could also introduce a `toHexString(BigInteger)` method in Debug. That will save you the extra `toByteArray()` call you're making in the various call sites.

It should be possible to extend code coverage in an existing x509 test also to ensure the serial number `toString `output is as expected and to ensure we don't regress in this area in the future etc.

**Review by seanjmullan — APPROVED** [review]
One comment on the test, but looks good otherwise.

**Review by coffeys — APPROVED** [review]
Looks good to me. 

Please ensure you run a wide selection of tests before integration. It should include JFR testing to exercise the X509 JFR event testing we have there.

**Pull request closed without merging**